### PR TITLE
fix[codegen]: panic on potential subscript eval order issue

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
panic on variant of https://github.com/vyperlang/vyper/pull/4157

### Commit message

```
subscript expressions have an evaluation order issue when
evaluation of the index (i.e. `node.index`) modifies the parent
(i.e. `node.value`). because the evaluation of the parent is
interleaved with evaluation of the index, it can result in "invalid"
reads where the length check occurs before evaluation of the index, but
the data read occurs afterwards. if evaluation of the index results in
modification of the container size for instance, the data read from the
container can happen on a dangling reference.

another variant of this issue would be accessing
`self.nested_array.pop().append(...)`; however, this currently happens
to be blocked by a panic in the frontend.

this commit conservatively blocks compilation if the preconditions for
the interleaved evaluation are detected. POC tests that the appropriate
panics are generated are included as well.

---------

Co-authored-by: trocher <trooocher@proton.me>
Co-authored-by: Hubert Ritzdorf <hubert.ritzdorf@chainsecurity.com>
Co-authored-by: cyberthirst <cyberthirst.eth@gmail.com>
```

### Cute Animal Picture

image

### CODE REVIEW

**Comment by cyberthirst** [comment]
passes
```python
def test_array_index_overlap_extcall(get_contract):
    code = """

interface Bar:
    def bar() -> uint256: payable    

a: public(DynArray[DynArray[Bytes[96], 5], 5])

@external
def foo() -> Bytes[96]:
    self.a.append([b'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'])
    return self.a[0][extcall Bar(self).bar()]


@external
def bar() -> uint256:
    self.a[0] = [b'yyy']
    self.a.pop()
    return 0
    """
    c = get_contract(code)
    assert c.foo() == b"yyy"
```

File: tests/functional/codegen/types/test_array_indexing.py [hunk 1]
```diff
@@ -130,3 +134,76 @@ def foo():
...
+
+@internal
+def bar() -> uint256:
+    self.a[0] = [b'yyy']
+    self.a.pop()
+    return 0
+    """
+    c = get_contract(code)
+    # tricky to get this right, for now we just panic instead of generating code
+    assert c.foo() == [b"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"]
```
Comment (cyberthirst) [inline tests/functional/codegen/types/test_array_indexing.py:159]: shouldn't be wrapped in `[ ]`, no?

File: tests/functional/codegen/types/test_array_indexing.py [hunk 2]
```diff
@@ -130,3 +134,76 @@ def foo():
...
+
+
+@external
+def bar() -> uint256:
+    self.a[0] = [b'yyy']
+    self.a.pop()
+    return 0
+    """
+    c = get_contract(code)
+    assert c.foo() == [b"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"]
```
Comment (cyberthirst) [inline tests/functional/codegen/types/test_array_indexing.py:185]: ditto

File: tests/functional/codegen/types/test_array_indexing.py (hunk 1 above)
Comment (charles-cooper) [reply tests/functional/codegen/types/test_array_indexing.py:159]: hmm true, but if we ever modify codegen so it passes `get_contract()`, the test needs to be fixed anyways

Comment (charles-cooper) [reply tests/functional/codegen/types/test_array_indexing.py]: but anyway, updated -- a397c73dd3bd9e30b566d82287fb52d4a51631e8

File: vyper/codegen/core.py [hunk 3]
```diff
@@ -924,6 +924,26 @@ def potential_overlap(left, right):
...
+    if not isinstance(left, IRnode) or not isinstance(right, IRnode):
+        return False
+
+    if left.typ._is_prim_word and right.typ._is_prim_word:
+        return False
+
+    if len(left.referenced_variables & right.variable_writes) > 0:
+        return True
+
+    if len(left.referenced_variables) > 0 and right.contains_risky_call:
```
Comment (cyberthirst) [inline vyper/codegen/core.py:941]: probably overly conservative as `contains_risky_call` checks against `staticcall`

**Review by cyberthirst — APPROVED** [review]

**Pull request merged**